format/luajit/luajit.go

```diff
@@ -0,0 +1,381 @@
+	if err != nil {
+		panic(err)
+	}
+
+	return f
+}
+
+type DumpInfo struct {
+	Strip     bool
+	BigEndian bool
+}
+
+func LuaJITDecodeHeader(di *DumpInfo, d *decode.D) {
+	d.FieldRawLen("magic", 3*8, d.AssertBitBuf([]byte{0x1b, 0x4c, 0x4a})) // ESC 'L' 'J'
+
+	d.FieldU8("version")
+
+	var flags uint64
+	d.FieldStruct("flags", func(d *decode.D) {
+		flags = d.FieldULEB128("raw")
+
+		d.FieldValueBool("be", flags&0x01 > 0)
+		d.FieldValueBool("strip", flags&0x02 > 0)
+		d.FieldValueBool("ffi", flags&0x04 > 0)
+		d.FieldValueBool("fr2", flags&0x08 > 0)
+	})
+
+	di.Strip = flags&0x2 > 0
+	di.BigEndian = flags&0x1 > 0
+
+	if !di.Strip {
+		namelen := d.FieldU8("namelen")
+		d.FieldUTF8("name", int(namelen))
+	}
+}
+
+type jumpBias struct{}
+
+func (j *jumpBias) MapUint(u scalar.Uint) (scalar.Uint, error) {
+	u.Actual -= 0x8000
+	return u, nil
+}
+
+func LuaJITDecodeBCIns(d *decode.D) {
+	op := d.FieldU8("op", opcodes)
+
+	d.FieldU8("a")
+
+	if opcodes[int(op)].HasD() {
+		if opcodes[int(op)].IsJump() {
+			d.FieldU16("j", &jumpBias{})
+		} else {
+			d.FieldU16("d")
+		}
+	} else {
+		d.FieldU8("c")
+		d.FieldU8("b")
+	}
+}
+
+type ktabType struct{}
+
+func (t *ktabType) MapUint(u scalar.Uint) (scalar.Uint, error) {
+	switch u.Actual {
+	case 0:
+		u.Sym = "nil"
+	case 1:
+		u.Sym = "false"
+	case 2:
+		u.Sym = "true"
+	case 3:
+		u.Sym = "int"
+	case 4:
+		u.Sym = "num"
+	default:
+		u.Sym = "str"
+	}
+	return u, nil
+}
+
+func LuaJITDecodeNum(d *decode.D) {
+	d.FieldAnyFn("num", func(d *decode.D) any {
+		lo := d.ULEB128()
+		hi := d.ULEB128()
+		return u64tof64((hi << 32) + lo)
+	})
+}
+
+func LuaJITDecodeKTabK(d *decode.D) {
+	ktabtype := d.FieldULEB128("ktabtype", &ktabType{})
+
+	if ktabtype >= 5 {
+		sz := ktabtype - 5
+		d.FieldUTF8("str", int(sz))
+	} else {
+		switch ktabtype {
+		case 3:
+			d.FieldULEB128("int")
+
+		case 4:
+			LuaJITDecodeNum(d)
+		}
+	}
+}
+
+func LuaJITDecodeCplx(d *decode.D) any {
+	lo := d.ULEB128()
+	if lo&1 == 0 {
+		return lo >> 1
+	} else {
+		hi := d.ULEB128()
+		return u64tof64((hi << 32) + (lo >> 1))
+	}
+}
+
+type kgcType struct{}
+
+func (t *kgcType) MapUint(u scalar.Uint) (scalar.Uint, error) {
+	switch u.Actual {
+	case 0:
+		u.Sym = "child"
+	case 1:
+		u.Sym = "tab"
+	case 2:
+		u.Sym = "i64"
+	case 3:
+		u.Sym = "u64"
+	case 4:
+		u.Sym = "complex"
+	default:
+		u.Sym = "str"
+	}
+	return u, nil
+}
+
+func LuaJITDecodeTab(d *decode.D) {
+	narray := d.FieldULEB128("narray")
+	nhash := d.FieldULEB128("nhash")
+
+	d.FieldArray("karray", func(d *decode.D) {
+		for i := uint64(0); i < narray; i++ {
+			d.FieldStruct("ktab", LuaJITDecodeKTabK)
+		}
+	})
+
+	d.FieldArray("khash", func(d *decode.D) {
+		for i := uint64(0); i < nhash; i++ {
+			d.FieldStruct("khash", func(d *decode.D) {
+				d.FieldStruct("k", LuaJITDecodeKTabK)
+				d.FieldStruct("v", LuaJITDecodeKTabK)
+			})
+		}
+	})
+}
+
+func LuaJITDecodeI64(d *decode.D) {
+	d.FieldAnyFn("i64", func(d *decode.D) any {
```

Comment:
For some other format decoders that do something similar i've decode the value as a "value" field, see msgpack or bson etc. Maybe that would make sense here also? for complex maybe the value field could an extra `d.FieldValueAny("value", ...)`?